packages/contracts/src/contracts/current/protocol/Exchange/MixinMatchOrders.sol

```diff
@@ -0,0 +1,285 @@
+        uint256 leftTakerAssetAmountRemaining = safeSub(leftOrder.takerAssetAmount, leftOrderFilledAmount);
+        if(safeMul(leftTakerAssetAmountRemaining, rightOrder.takerAssetAmount) <= safeMul(rightTakerAssetAmountRemaining, rightOrder.makerAssetAmount))
+        {
+            // Left order is the constraint: maximally fill left
+            (   status,
+                matchedFillOrderAmounts.left
+            ) = calculateFillAmounts(
+                leftOrder,
+                leftOrderStatus,
+                leftOrderFilledAmount,
+                leftTakerAssetAmountRemaining);
+            if(status != uint8(Status.SUCCESS)) {
+                return (status, matchedFillOrderAmounts);
+            }
+
+            // The right order just spent <leftTakerAssetAmountRemaining> of their maker asset to fill the left order.
+            // The amount right gets in return is:
+            //    <leftOrderAmountBought> * <rightOrderProfitPerUnitSold>
+            // =  <matchedFillOrderAmounts.left.takerAssetFilledAmount> * <rightOrder.takerAssetAmount> / <rightOrder.makerAssetAmount>
+            // TODO: Ensure rounding error is in the right direction.
+            if(isRoundingError(rightOrder.takerAssetAmount, rightOrder.makerAssetAmount, matchedFillOrderAmounts.left.takerAssetFilledAmount)) {
+                status = uint8(Status.ROUNDING_ERROR_TOO_LARGE);
+                return (status, matchedFillOrderAmounts);
+            }
+            uint256 rightOrderAmountToFill = getPartialAmount(
+                rightOrder.takerAssetAmount,
+                rightOrder.makerAssetAmount,
+                matchedFillOrderAmounts.left.takerAssetFilledAmount);
+
+            // Compute fill amounts for right order
+            (   status,
+                matchedFillOrderAmounts.right
+            ) = calculateFillAmounts(
+                rightOrder,
+                rightOrderStatus,
+                rightOrderFilledAmount,
+                rightOrderAmountToFill);
+            if(status != uint8(Status.SUCCESS)) {
+                return (status, matchedFillOrderAmounts);
+            }
+
+            // The right order must spend at least as much as we're transferring to the left order's maker.
+            // If the amount transferred from the right order is greater than what is transferred, it is a rounding error amount.
+            // Ensure this difference is negligible by dividing the values with each other. The result should equal to ~1.
+            assert(matchedFillOrderAmounts.right.makerAssetFilledAmount >= matchedFillOrderAmounts.left.takerAssetFilledAmount);
+            if(isRoundingError(matchedFillOrderAmounts.right.makerAssetFilledAmount, matchedFillOrderAmounts.left.takerAssetFilledAmount, 1)) {
+                status = uint8(Status.ROUNDING_ERROR_TOO_LARGE);
+                return (status, matchedFillOrderAmounts);
+            }
+        } else {
```

Comment:
There's a lot of duplicate logic in this else block. Can we break this out into another function?

Reply:
Good call 👍

Reply:
Updated